controllers/networkfence_controller.go

```diff
@@ -18,38 +18,135 @@ package controllers
 )
 
-// NetworkFenceReconciler reconciles a NetworkFence object
+// NetworkFenceReconciler reconciles a NetworkFence object.
 type NetworkFenceReconciler struct {
 	client.Client
-	Scheme *runtime.Scheme
+	Scheme   *runtime.Scheme
+	Connpool *conn.ConnectionPool
 }
 
+var (
+	networkFenceFinalizer = csiaddonsv1alpha1.GroupVersion.Group + "/networkfence"
+)
+
 //+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=networkfences,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=networkfences/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=networkfences/finalizers,verbs=update
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the NetworkFence object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
 func (r *NetworkFenceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// fetch NetworkFence object instance
+	nwFence := &csiaddonsv1alpha1.NetworkFence{}
+	err := r.Get(ctx, req.NamespacedName, nwFence)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("NetworkFence resource not found")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// validate NetworkFence object so as its parameters are neither empty nor nil.
+	err = validateNetworkFenceSpec(nwFence)
+	if err != nil {
+		logger.Error(err, "failed to validate NetworkFence spec")
+
+		errMessage := util.GetErrorMessage(err)
+		nwFence.Status.Result = csiaddonsv1alpha1.FencingOperationResultFailed
+		nwFence.Status.Message = fmt.Sprintf("Failed to validate Networkfence parameters: %v", errMessage)
+		statusErr := r.Client.Status().Update(ctx, nwFence)
+		if statusErr != nil {
+			logger.Error(statusErr, "Failed to update status")
+
+			return ctrl.Result{}, statusErr
+		}
+
+		// invalid parameter, do not requeue
+		return ctrl.Result{}, nil
+	}
+
+	logger = logger.WithValues("Driver Name", nwFence.Spec.Driver, "CIDRs", nwFence.Spec.Cidrs)
 
-	// TODO(user): your logic here
+	conns, unlock := r.Connpool.Get()
+	var (
+		controllerClient *grpc.ClientConn
+	)
+	for _, v := range *conns {
+		if v.DriverName == nwFence.Spec.Driver {
+			for _, cap := range v.Capabilities {
+				if cap.GetNetworkFence() == nil {
+					continue
+				}
+				// check if the object is being deleted
+				if nwFence.GetDeletionTimestamp().IsZero() {
```

Comment:
this can be structured simpler. Do an error check and return inside the `if`, when no error happens, continue the function normally (no `else` needed).